server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala

```diff
@@ -12,131 +14,126 @@
+    userProgressOps: UserProgressOps[IO],
+    userExercisesProgress: UserExercisesProgress[IO],
+    githubOps: GithubOps[IO],
+    mode: Mode)
+    extends BaseController
+    with AuthenticationModule {
+  lazy val topLibraries: List[String] = config.getOptional[Seq[String]]("exercises.top_libraries") map (_.toList) getOrElse Nil
 
   val MainRepoCacheKey = "scala-exercises.repo"
 
+  val configUtils = ConfigUtils(config)
+
+  private val service = new ExercisesService(cl)
+
+  private implicit val cs = IO.contextShift(defaultExecutionContext)
+
   /** cache the main repo stars, forks and watchers info for 30 mins */
-  private[this] def scalaexercisesRepo: FreeS[ExercisesApp.Op, Repository] = {
-    cache.get[Repository](MainRepoCacheKey) match {
-      case Some(repo) ⇒ FreeS.pure(repo)
+  private[this] def scalaexercisesRepo: IO[Option[Repository]] = {
+    IO.fromFuture(IO(cache.get[Repository](MainRepoCacheKey)))(cs).flatMap {
+      case repo if repo.nonEmpty ⇒ IO.pure(repo)
       case None ⇒
         githubOps
           .getRepository(
-            ConfigUtils.githubSiteOwner,
-            ConfigUtils.githubSiteRepo
-          ) match {
-          case repo ⇒
-            cache.set(MainRepoCacheKey, repo, 30 minutes)
-            repo
-          case err ⇒
-            Logger.error("Error fetching scala-exercises repository information")
-            err
-        }
+            configUtils.githubSiteOwner,
+            configUtils.githubSiteRepo
+          )
+          .redeem({ _ =>
+            Logger(this.getClass).error("Error fetching scala-exercises repository information")
+            None
+          }, repo => Some(repo))
     }
   }
 
   def index =
-    Secure(Action.async { implicit request ⇒
-      for {
-        authorize ← githubOps.getAuthorizeUrl(ConfigUtils.githubAuthId, ConfigUtils.callbackUrl)
-        libraries ← exerciseOps.getLibraries.map(ExercisesService.reorderLibraries(topLibraries, _))
+    Secure(Action.async { implicit request =>
+      (for {
+        authorize ← githubOps
+          .getAuthorizeUrl(configUtils.githubAuthId, configUtils.callbackUrl)
+        libraries ← exerciseOps.getLibraries.map(service.reorderLibraries(topLibraries, _))
         user      ← userOps.getUserByLogin(request.session.get("user").getOrElse(""))
         progress  ← userExercisesProgress.fetchMaybeUserProgress(user)
         repo      ← scalaexercisesRepo
         result = (libraries, user, request.session.get("oauth-token"), progress, authorize) match {
-          case (libraries, user, Some(token), progress, _) ⇒
+          case (libraries, user, Some(_), progress, _) ⇒
             Ok(
               views.html.templates.home
-                .index(user = user, libraries = libraries, progress = progress, repo = repo))
+                .index(
+                  user = user,
+                  libraries = libraries,
+                  progress = progress,
+                  repo = repo.getOrElse(Repository(0, 0, 0)),
```

Comment:
0, 0, 0? what is that?

Reply:
Is a default value for the numbers of Scala Exercises repo shown at the index if connection with GitHub fails. Any alternative?